[DataGrid] Add a subset of used `@mui/material` components to `GridSlotsComponent`

DanailH opened the pull request:
Part of #3066 

I've made a subsite of all `@mui/material` components we use in the grid be available as a slot component. I've checked all the `@mui/material` we use and these once seemed to make the most sense for a first iteration. I'm adding the rest below. We can discuss and see if any of the others also need to be added in this first iteration (or if I need to revert any of the ones I added).

---

**List of remaining `@mui/material` components currently used in the grid:**
- CircularProgress
- TablePagination
- MenuList
- MenuItem
- Paper
- Grow
- Popper
- IconButton
- ListItemIcon
- InputBase
- InputLabel
- FormControlLabel
- Box
- Badge
- Tooltip
- NoSsr
- ClickAwayListener
- Grow
- MuiDivider

flaviendelangle commented:
Could we add a `Form input` section [here](https://mui.com/components/data-grid/components/#components)

DanailH commented:
> Could we add a `Form input` section [here](https://mui.com/components/data-grid/components/#components) ? To show how to use it to switch to another variant for instance.

Sure, but I didn't add all the form elements we use from the core. I skipped `InputBase`, `InputLabel`, `FormControlLabel`. If we are adding a `Form input` section in the docs maybe its better to also add slots for those components before that?

flaviendelangle commented:
Yes, if you prefer we can first release this feature in a discrete way and then document it fully when ready.

m4theushw commented on packages/grid/_modules_/grid/components/panel/filterPanel/GridFilterPanel.tsx:
```diff
@@ -101,9 +100,13 @@ export function GridFilterPanel() {
+          <rootProps.components.BaseButton
+            onClick={addNewFilter}
+            startIcon={<GridAddIcon />}
+            color="primary"
```
```suggestion
            color="primary"
            {...rootProps.componentsProps?.baseButton}
```

x other slots.

DanailH replied:
Added

m4theushw reviewed (commented):
Should `Popper` also be added? Adding it seems to solve part of the problem raised in #3381.

> It is possible to override Portal props for Column Header tooltips by customizing the theme's default props for Tooltip, but since there is no ability to customize default props for Popper/Portal to propogate the shadow dom container, all of the styled(Popper) components will appear outside and have no styling applied.

DanailH commented:
> Should `Popper` also be added? Adding it seems to solve part of the problem raised in #3381.
> 
> > It is possible to override Portal props for Column Header tooltips by customizing the theme's default props for Tooltip, but since there is no ability to customize default props for Popper/Portal to propogate the shadow dom container, all of the styled(Popper) components will appear outside and have no styling applied.

I added `componentsProps` for `Popper` since we never use it directly. We use it as a base for `GridMenu`. In addition, I added the `Tooltip` to slot components.

m4theushw commented on packages/grid/_modules_/grid/models/gridSlotsComponent.ts:
```diff
@@ -6,15 +6,45 @@ import { GridIconSlotsComponent } from './gridIconSlotsComponent';
+   */
+  BaseTooltip: React.JSXElementConstructor<any>;
+  /**
+   * Component rendered for each cell.
```
```suggestion
   * Component rendered for each cell.
   * @default GridCell
```

m4theushw commented on packages/grid/_modules_/grid/components/menu/GridMenu.tsx:
```diff
@@ -98,6 +98,7 @@ const GridMenu = (props: GridMenuProps) => {
       transition
```
Could you add the slot for `Popper`? To keep each `components` key with a matching `componentsProps` key. I guess you need to pass the component using the `as` prop.

DanailH replied:
Done

m4theushw reviewed (commented):
It's missing to spread `rootProps.componentsProps?.basePopper` in the `GridPanel`.

https://github.com/mui-org/material-ui-x/blob/7bac90d6dbaef58457fda74efa224d7d97dd8f4e/packages/grid/_modules_/grid/components/panel/GridPanel.tsx#L30

m4theushw commented on packages/grid/_modules_/grid/components/panel/GridPanel.tsx:
```diff
@@ -23,11 +23,12 @@ export interface GridPanelProps extends StandardProps<PopperProps, 'children'> {
 export const gridPanelClasses = generateUtilityClasses('MuiDataGrid', ['panel', 'paper']);
 
-const GridPanelRoot = styled(Popper, {
+const GridPanelRoot = styled('div', {
```
Although I suggested it, I don't know if this is the right approach. Looking at the change done in https://github.com/mui-org/material-ui-x/pull/3490/files#diff-c2fd479a822e3ad62e4338f61294f31a114e307bcbcce46c482822ef9d8fc784 maybe the default tag should be `Popper` and keep the `as` prop too. 

Looking closer I found a similar case in the core:

https://github.com/mui-org/material-ui/blob/bc016a59f8b4afeb43c98f677c9fbd1e8afb2698/packages/mui-material/src/Autocomplete/Autocomplete.js#L408

https://github.com/mui-org/material-ui/blob/bc016a59f8b4afeb43c98f677c9fbd1e8afb2698/packages/mui-material/src/Autocomplete/Autocomplete.js#L581-L582

The default value of `PopperComponent` is `Popper`.

DanailH replied:
Updated

m4theushw commented on packages/grid/_modules_/grid/components/panel/GridPanel.tsx:
```diff
@@ -23,11 +23,12 @@ export interface GridPanelProps extends StandardProps<PopperProps, 'children'> {
    */
   classes?: Partial<GridPanelClasses>;
   open: boolean;
+  as: React.ElementType;
```
You can extend from `MUIStyledCommonProps` as below:

```diff
diff --git a/packages/grid/_modules_/grid/components/panel/GridPanel.tsx b/packages/grid/_modules_/grid/components/panel/GridPanel.tsx
index 510e869a2..68f20b1a7 100644
--- a/packages/grid/_modules_/grid/components/panel/GridPanel.tsx
+++ b/packages/grid/_modules_/grid/components/panel/GridPanel.tsx
@@ -1,8 +1,9 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
 import clsx from 'clsx';
-import { styled } from '@mui/material/styles';
+import { styled, Theme } from '@mui/material/styles';
 import { generateUtilityClasses, InternalStandardProps as StandardProps } from '@mui/material';
+import { MUIStyledCommonProps } from '@mui/system';
 import ClickAwayListener from '@mui/material/ClickAwayListener';
 import Paper from '@mui/material/Paper';
 import { PopperProps } from '@mui/material/Popper';
@@ -16,14 +17,14 @@ export interface GridPanelClasses {
   paper: string;
 }
 
-export interface GridPanelProps extends StandardProps<PopperProps, 'children'> {
+export interface GridPanelProps
+  extends StandardProps<MUIStyledCommonProps<Theme> & PopperProps, 'children'> {
   children?: React.ReactNode;
   /**
    * Override or extend the styles applied to the component.
    */
   classes?: Partial<GridPanelClasses>;
   open: boolean;
-  as: React.ElementType;
 }
 
 export const gridPanelClasses = generateUtilityClasses('MuiDataGrid', ['panel', 'paper']);
```

m4theushw commented on packages/grid/_modules_/grid/components/panel/GridPanel.tsx:
```diff
@@ -76,7 +78,6 @@ const GridPanel = React.forwardRef<HTMLDivElement, GridPanelProps>((props, ref)
       ref={ref}
       placement="bottom-start"
       className={clsx(className, classes.panel)}
```
The `as` prop is missing as well as `rootProps.componentsProps.panel`.

DanailH replied:
I can't add them in the `GridPanel` directly because the `useGridRootProps` can be used in components outside of `DataGrid`. That's why I had to pass it down from the parent. The same applies to the other comment https://github.com/mui-org/material-ui-x/pull/3490#discussion_r779507845

```
Error: Uncaught Error: MUI: useGridRootProps should only be used inside the DataGrid/DataGridPro component.

m4theushw replied:
We had a discussion about that in #2522.

Currently the `GridPanel` already depends on another context: `GridApiContext`. We could also use `useGridRootProps` but that would be a breaking change. So I'm OK to not do that for now.

DanailH replied:
Ok so if we are postponing this and if this is the only thing left can we proceed with merging this PR?

m4theushw commented on packages/grid/_modules_/grid/components/panel/GridPreferencesPanel.tsx:
```diff
@@ -22,9 +22,11 @@ export const GridPreferencesPanel = React.forwardRef<
   return (
     <rootProps.components.Panel
       ref={ref}
+      as={rootProps.components.BasePopper}
       open={columns.length > 0 && preferencePanelState.open}
       {...rootProps.componentsProps?.panel}
```
If the `as` prop is applied inside `GridPanel`, then it doesn't need to be applied here again.
